cleverhans/confidence_report.py

```diff
@@ -0,0 +1,129 @@
+"""Functionality for making confidence reports.
+
+A confidence report is a dictionary.
+Each dictionary key is the name of a type of data:
+  clean : Clean data
+  bundled : bundled adversarial examples
+Each value in the dictionary contains an array of bools indicating whether
+the model got each example correct and an array containing the confidence
+that the model assigned to each prediction.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import logging
+
+import numpy as np
+import tensorflow as tf
+
+from cleverhans.utils import set_log_level
+from cleverhans.serial import load
+
+# Defaults. Imported elsewhere so that command line script defaults match
+# function defaults.
+TRAIN_START = 0
+TRAIN_END = 60000
+TEST_START = 0
+TEST_END = 10000
+WHICH_SET = 'test'
+
+def make_confidence_report_bundled(filepath, train_start=TRAIN_START,
+                                   train_end=TRAIN_END, test_start=TEST_START,
+                                   test_end=TEST_END, which_set=WHICH_SET):
+  """
+  Load a saved model, gather its predictions, and save a confidence report.
+  :param filepath: path to model to evaluate
+  :param train_start: index of first training set example to use
+  :param train_end: index of last training set example to use
+  :param test_start: index of first test set example to use
+  :param test_end: index of last test set example to use
+  :param which_set: 'train' or 'test'
+  """
+  # Avoid circular import
+  from cleverhans.attack_bundling import basic_max_confidence_recipe
+
+  # Set TF random seed to improve reproducibility
+  tf.set_random_seed(1234)
+
+  # Set logging level to see debug information
+  set_log_level(logging.INFO)
+
+  # Create TF session
+  sess = tf.Session()
+
+  assert filepath.endswith('.joblib')
+  report_path = filepath[:-len('.joblib')] + "_bundled_report.joblib"
+
+  with sess.as_default():
+    model = load(filepath)
+  assert len(model.get_params()) > 0
+  factory = model.dataset_factory
+  factory.kwargs['test_start'] = test_start
+  factory.kwargs['test_end'] = test_end
+  dataset = factory()
+
+  center = factory.kwargs['center']
+  value_range = 1. + center
+  min_value = 0. - center
+  max_value = 1.
+
+  if 'CIFAR' in str(factory.cls):
```

Comment:
Would it make sense to pass these as arguments rather than hardcoding them here so this can be used with other datasets?

Reply:
At the moment, these really are just hard-coded values that people use for each dataset in published papers. I don't know a way of auto-inferring them for new datasets.

Reply:
Later we might add some kind of config system. I'd prefer to save that for a later PR though.

Reply:
Ah I see why the values were hard-coded then. Would it make sense to move this to scripts or do you think it's a feature we want people to be able to import directly from the CleverHans module?